Support running command-line specified list of tests.

eengstrom opened the pull request:
Add support for command-line specification of tests to run.

This works because `shunit2` previously assumes/accepts only one argument, which if provided is the name of the file that contains the tests. This modification takes all additional arguments to be the names of the test functions to be run.

By example, will support either:
    ./test-script.sh -- test1 ... testN
or
    shunit2 test-script.sh test1 ... testN

The first case requires an additional no-op argument ('--') to tell `shunit2` that additional test arguments are coming.

Resolves #23

kward commented on shunit2:
```diff
@@ -1140,11 +1140,23 @@ oneTimeSetUp
+  # arguments $2 .. $# are assumed to be test function names
+  # interate through all remaining args in "$@" in a POSIX (likely portable) way
+  # helpful tip: https://unix.stackexchange.com/questions/314032/how-to-use-arguments-like-1-2-in-a-for-loop
+  for arg_test_ do
```
The arg_test_ variable doesn't fit the variable name standard for shunit2.
https://github.com/kward/shunit2/wiki/Coding-standards

```suggestion
  for shunit_arg_ do
```

eengstrom replied:
Updated to follow coding conventions; see commit @6f8bc7a.

kward commented on shunit2:
```diff
@@ -1140,11 +1140,23 @@ oneTimeSetUp
-suite
+# Command line selected tests or suite selected tests
+if command [ "$#" -ge 2 ]; then
+  # arguemnt $1 is either the filename of tests or '--'; either way, skip it.
```
Could you make the comments full sentences, including capitalization?

kward commented on shunit2_args_test.sh:
```diff
@@ -0,0 +1,46 @@
+# Load and run shunit2.
+# shellcheck disable=SC2034
+[ -n "${ZSH_VERSION:-}" ] && SHUNIT_PARENT=$0
+# if zero/one arguments provided, artificially set the arguments to this script, in a POSIX way
```
As you are trying to demonstrate the new functionality of passing arguments, I don't think it makes sense to make that functionality conditional (line 42). Replace this comment with a statement of what you are doing in line 43, and remove the if/fi in 42/44.

eengstrom replied:
Actually, I intended that conditional to allow one to test the test script itself by using command-line arguments to specify individual tests.  In that case, I wouldn't want to override the user-specified tests.  Therefore, I improved the comments (see commit 298c99ef882ff09da6c99349c6442cd6ce63bb71) to better explain what I did rather than change the functionality.  Thoughts?

eengstrom commented:
So, I think I've addressed your concerns by improving the commentary in several spots, as well as adhering to the coding conventions.  I also found a latent issue for zsh that I fixed in 75f2ee41c099e8f4c8058f5bc90a79dfedeee3c9.  Finally, I also fixed (reverted) a very recent change you made that broke `assertNotEquals` when given message arguments.  

Merged your current Master to make diffs and merging easier for you and to ensure I'm not introducing any new errors.

I hope this is helpful.

eengstrom commented on shunit2:
```diff
@@ -1285,11 +1285,24 @@ oneTimeSetUp
 
-# If no suite function was defined, dynamically build a list of functions.
+# If no tests specified defined, dynamically build a list of functions.
 if command [ -z "${__shunit_suite}" ]; then
```
Not sure I follow what you are saying.  As it stands (with my changes), you can run a test script either:

1. without arguments, which implicitly runs all things named `test*`, or the test suite if one exists, or
2. with arguments, e.g. `test_script.sh test_foo test_bar`, in which case only those (two) tests will run.

Are you saying if both `test_foo` and `test_bar` do not exist, NO tests will run?  That's sort of the case, as it will actually try to run those functions and fail as they don't exist.  My changes do no checking to see if the test functions specified exist.  

Is there a third option I'm not seeing?

cbandy replied:
That sounds good. I misunderstood the code on this line and was reacting to the comment

> If no tests specified defined,

Is "defined" there a typo?

eengstrom replied:
Yeah, that's a residual typo - fixed in 584f14636e6f08cb1430499486657fa5ba06404d.

Also added documentation to the README.md that I should have written the first time around.

kward commented:
Thank you for the contribution!

stephenmoloney commented:
I am using shunit2 and I was trying to figure out why my tests were failing and this change is the cause.

I am passing custom into a shunit2 script but this commit makes it no longer possible to do so.

What do you think of the idea of specifying the tests with commandline arguments, eg `--tests="test1 test2 test3"` and then the remaining arguments picked up as normal.

Any thoughts ?

eengstrom commented:
Since I was the author of the changes that Kate kindly merged, I thought I'd respond.

> I am passing custom arguments into a shunit2 script but this commit makes it no longer possible to do so. I do like the possibility of cherry picking tests though.

I've not used additional arguments to `shunit` scripts in this way.  Are you processing the arguments before you source `shunit`?  Can you provide an example command line invocation?  Better yet, a (sanitized) example script that makes use of them?  I ask because you are using `shunit` very differently from my typical usages, and if I (arbitrarily) add any command line arguments, `shunit` assumes that the arguments are files to be read and therefore raises a fatal error.  Therefore, I'm not sure what arguments you want to pass to `shunit`.

> What do you think of the idea of specifying the tests with command line arguments, eg `--only="test1 test2 test3"` and then the remaining arguments picked up as normal.

That might work.  However, regardless of the form of the arguments, your test script would have to pass on to (leave in `"$@"`) `shunit` any arguments you didn't want for yourself.  Or perhaps I'm completely missing a use case, which is entirely likely.

stephenmoloney commented:
@eengstrom 

Thanks for the reply.

I am not sourcing shunit2 but rather calling it directly with arguments
Here is an illustration of how I am currently passing arguments to the shunit2 script:

```
/tmp/shunit2/shunit2 "${PWD%/*}/test/shellcheck_tests.sh" \
  "${IMAGE}" \
  "${RELATIVE_PATH}" \
  "${HOST_VOLUME_PATH}" \
  "${SHELLCHECK_VERSION}" \
  "${SKIP_TEAR_DOWN}"
```

extracted from this file:

https://github.com/stephenmoloney/ci-actions-shellcheck/blob/2bdf89ad4cc33b9659d81a3d50de038c712628ee/shellcheck/hooks/test#L20-L25
